cupy/core/_routines_indexing.pyx

```diff
@@ -0,0 +1,819 @@
+        return _take_kernel_scalar(
+            a.reduced_view(), indices, ldim, cdim, rdim, index_range, out)
+
+
+cdef _scatter_op_single(
+        ndarray a, ndarray indices, v, Py_ssize_t li=0, Py_ssize_t ri=0,
+        op=''):
+    # When op == 'update', this function behaves similarly to
+    # a code below using NumPy under the condition that a = a._reshape(shape)
+    # does not invoke copy.
+    #
+    # shape = a[:li] +\
+    #     (numpy.prod(a[li:ri+1]),) + a[ri+1:]
+    # a = a._reshape(shape)
+    # slices = (slice(None),) * li + indices +\
+    #     (slice(None),) * (a.ndim - indices.ndim - ri)
+    # a[slices] = v
+    cdef Py_ssize_t ndim, adim, cdim, rdim
+    cdef tuple a_shape, indices_shape, lshape, rshape, v_shape
+
+    ndim = a._shape.size()
+
+    if ndim == 0:
+        raise ValueError("requires a.ndim >= 1")
+    if not (-ndim <= li < ndim and -ndim <= ri < ndim):
+        raise ValueError('Axis overrun')
+
+    if not isinstance(v, ndarray):
+        v = core.array(v, dtype=a.dtype)
+    else:
+        v = v.astype(a.dtype, copy=False)
+
+    a_shape = a.shape
+    li %= ndim
+    ri %= ndim
+
+    lshape = a_shape[:li]
+    rshape = a_shape[ri + 1:]
+    adim = internal.prod(a_shape[li:ri + 1])
+
+    indices_shape = indices.shape
+    v_shape = lshape + indices_shape + rshape
+    v = _manipulation.broadcast_to(v, v_shape)
+
+    cdim = indices.size
+    rdim = internal.prod(rshape)
+    indices = _manipulation._reshape(
+        indices,
+        (1,) * len(lshape) + indices_shape + (1,) * len(rshape))
+    indices = _manipulation.broadcast_to(indices, v_shape)
+
+    if op == 'update':
+        _scatter_update_kernel(
+            v, indices, cdim, rdim, adim, a.reduced_view())
+    elif op == 'add':
+        # There is constraints on types because atomicAdd() in CUDA 7.5
+        # only supports int32, uint32, uint64, and float32.
+        if not issubclass(v.dtype.type,
+                          (numpy.int32, numpy.float16, numpy.float32,
+                           numpy.float64, numpy.uint32, numpy.uint64,
+                           numpy.ulonglong)):
+            raise TypeError(
+                'scatter_add only supports int32, float16, float32, float64, '
+                'uint32, uint64, as data type')
+        _scatter_add_kernel(
+            v, indices, cdim, rdim, adim, a.reduced_view())
+    else:
+        raise ValueError('provided op is not supported')
+
+
+cdef _scatter_op_mask_single(ndarray a, ndarray mask, v, Py_ssize_t axis, op):
+    cdef ndarray mask_scanned, src
+    cdef tuple masked_shape
+
+    mask, mask_scanned, masked_shape = _prepare_mask_indexing_single(
+        a, mask, axis)
+    if internal.prod(masked_shape) == 0:
+        return
+
+    if not isinstance(v, ndarray):
+        src = core.array(v, dtype=a.dtype)
+    else:
+        src = v
+        # Cython's static resolution does not work because of omitted arguments
+        src = (<object>src).astype(a.dtype, copy=False)
```

Comment:
Added cast to `object`. See the comment for the reason.

Reply:
Previously Cython did this automatically, but it would complain now for some reason:
```
            src = src.astype(a.dtype, copy=False)
                           ^
cupy/core/_routines_indexing.pyx:620:24: C function call is missing argument 'order'
```